primitives/runtime/src/offchain/rpc.rs

```diff
@@ -0,0 +1,353 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A high-level helpers for making RPC calls from Offchain Workers.
+//!
+//! Example:
+//! ```rust,no_run
+//! use sp_runtime::offchain::rpc::{Request, Error};
+//! use sp_runtime::offchain::http;
+//! use sp_std::{vec::Vec, str};
+//!
+//! let request = Request {
+//! 	jsonrpc: "2.0",
+//! 	id: 1,
+//! 	method: "chain_getFinalizedHead",
+//! 	params: Vec::new(),
+//! 	timeout: 3_000,
+//! 	url: "http://localhost:9933"
+//! };
+//!
+//! let rpc_response = request.send();
+//!
+//! match rpc_response {
+//! 	Ok(response) => {
+//! 		if !response.result.is_empty() {
+//! 			log::info!("Rpc call result: {:?}", str::from_utf8(&response.result).unwrap());
+//! 		} else {
+//! 			log::error!("Rpc call error: code: {:?}, message: {:?}",
+//! 							response.error.code,
+//! 							str::from_utf8(&response.error.message).unwrap()
+//! 						);
+//! 		}
+//! 	},
+//! 	Err(e) => {
+//! 		log::error!("Rpc call error: {:?}", e);
+//! 		assert_eq!(Error::Http(http::Error::DeadlineReached), e);
+//! 	}
+//! }
+//! ```
+
+use serde::{Deserialize, Deserializer};
+use crate::{RuntimeDebug, offchain::Duration};
+use serde_json::json;
+use sp_std::{vec::Vec, str};
+use super::http;
+
+/// A rpc call error
+#[derive(Clone, PartialEq, Eq, RuntimeDebug)]
+pub enum Error {
+	/// Response can not be deserialized
+	Deserializing,
+	/// Http error
+	Http(http::Error),
+}
+
+/// A rpc result type
+pub type RpcResult = Result<Response, Error>;
+
+/// A rpc call error field
+#[derive(Deserialize, Default, RuntimeDebug)]
+pub struct RpcError {
+	/// error code
+	pub code: i32,
+	/// error message
+	#[serde(deserialize_with = "de_string_to_bytes")]
+	pub message: Vec<u8>,
+}
+
+/// A rpc call response
+#[derive(Deserialize, Default, RuntimeDebug)]
+pub struct Response {
+	/// rpc response jsonrpc
+	#[serde(deserialize_with = "de_string_to_bytes")]
+	pub jsonrpc: Vec<u8>,
+	/// rpc response result
+	#[serde(default)]
+	#[serde(deserialize_with = "de_string_to_bytes")]
+	pub result: Vec<u8>,
+	/// rpc response error struct
+	#[serde(default)]
+	pub error: RpcError,
+	/// rpc response id
+	pub id: u32,
+}
+
+/// A rpc call request
+pub struct Request<'a> {
+	/// rpc request jsonrpc
+	pub jsonrpc: &'a str,
+	/// rpc request id
+	pub id: u32,
+	/// rpc request method
+	pub method: &'a str,
+	/// rpc request params
+	pub params: Vec<&'a str>,
+	/// rpc request timeout
+	pub timeout: u64,
+	/// rpc request url
+	pub url: &'a str,
+}
+
+fn de_string_to_bytes<'de, D>(de: D) -> Result<Vec<u8>, D::Error>
+where
+	D: Deserializer<'de>,
+{
+	let s: &str = Deserialize::deserialize(de)?;
+	Ok(s.as_bytes().to_vec())
+}
+
+impl<'a> Request<'a> {
+	/// Send the request and return a RPC result.
+	///
+	/// Err is returned in case there is a Http
+	/// or deserializing error.
+	pub fn send(&self) -> RpcResult {
```

Comment:
```suggestion
	pub fn send(&self, url: &str) -> RpcResult {
```

To avoid mixing `Request` body with particular url.

Reply:
Now I think about this a bit further there are couple of things that lead me to proposing separation of `Request` and `Client` objects.

1. The URL will most likely be the same for mutliple requests.
2. Since the low-level HTTP api supports deadlines I'd prefer to keep this API in line.
3. But I do understand that for many RPC cases it might be easiser to think in timeouts instead of deadlines,

So as a consequence of this points I'd like to propose something like this:
```rust
#[derive(RuntimeDebug, Clone)]
pub struct Client {
 pub url: &str,
 pub request_timeout: u64,
}

impl Client {
  pub fn send(&self, request: &Request) -> ...{
    let deadline = sp_io::offchain::timestamp().add(Duration::from_millis(self.request_timeout));
    self.send_with_deadline(request, deadline);
  }
  pub fn send_with_deadline(&self, request: &Request, deadline: Timestamp) -> ... {
    /// rest of the sending code
  }
}
```
This gives us flexibility of using either timeouts or deadlines, and decouples Request from actual sending logic.

Reply:
- I think that adding `Default` implementation, plus `Clone` trait, plus modifiers methods (`fn url(mut self, url: &'a str)`, `fn timeout(mut self, timeout: u64)`, etc) should be sufficient to allow creating different custom request if needed.

  ```rust
  let request = rpc::Request::new(); // defaults to "http://localhost:9933" and 3_000
  
  let request_2 = request.clone().url("http://my_url").timeout(5_000);
  ```
  I'd like to follow the same approach as for `http` module https://github.com/paritytech/substrate/blob/master/primitives/runtime/src/offchain/http.rs#L132 
  where there is not really a need for adding a new `Client` struct.

- I modified `send` and added `send_with_deadline()` methods

Reply:
The reason I want to decouple `Request` and using HTTP transport is exactly the issue that is linked in description #6742
I expect that soonish it won't be necessary to send requests to an external URL, but rather we will submit them internally via `sp_io` host function. With this in place it will be awesome to re-use existing code to compose (serialize) the request, but `url` will not be necessary, hence to avoid breaking compatibility in the future I'd prefer to keep this two things separate.

As a compromise maybe we can do this via an extension trait (or just plain function):
```rust
impl HttpJsonRpc for Request<> {
  fn send(&self, url: &str) -> ... {
     ...
  }
}
// and in the future
impl InternalJsonRpc for Request<> {
  fn send(&self) -> ... { // no url
     ...
  }
}
```

TBH I also find passing `url` with every request a bit unergonomic, hence suggested the client instead of `send` function.

Reply:
Ok, understood, so I decoupled `Client` and `Request`, so with this change, if in the future internal RPC calls are implemented, `Request` can still be used independently for that purpose. Thus, your last suggestion extending `Request` with two different traits does not apply anymore.

Now, your previously suggested code is implemented:

```rust
#[derive(RuntimeDebug, Clone)]
pub struct Client {
 pub url: &str,
 pub request_timeout: u64,
}

impl Client {
  pub fn send(&self, request: &Request) -> ...{
    let deadline = sp_io::offchain::timestamp().add(Duration::from_millis(self.request_timeout));
    self.send_with_deadline(request, deadline);
  }
  pub fn send_with_deadline(&self, request: &Request, deadline: Timestamp) -> ... {
    /// rest of the sending code
  }
}
```

Even though this new `rpc` module was created to help with #6742 (to know the finalized block), it can be seen as a new module to perform RPC calls using HTTP transport. It can be used for any external service, or for example to query some other Substrate node.

Please, let me know if something else is still missing so this PR can be finally be approved :)